oeps/oep-0004.rst

```diff
@@ -0,0 +1,177 @@
+
+Specification
+=============
+
+Data model
+----------
+
+============ ==================================================================
+Object       Description
+============ ==================================================================
+service user an Open edX account created solely for server-to-server
+             integrations; can have multiple associated applications
+application  an autonomous system that needs to integrate with an Open edX
+             platform, owned by a single service user
+resource     named grouping of related capabilities. Will generally (not
+             always) be a noun like "catalog"
+action       describes a high-level behavior and will generally be "read" or
+             "write", but resources may use non-standard definitions where
+             absolutely necessary (should be the exception)
+scope        a single authorization (naming convention: ``resource:action``)
+             for an application relative to an Open edX instance; should only
+             describe broad application authorization (as opposed to, say,
+             fine-grained user authorizations)
+access token a temporary credential that grants a set of scopes to an
+             application
+============ ==================================================================
+
+I sketched these relationships out in the diagram below. Note that instead of using "read/write" as actions, I noted the HTTP verbs instead, just to give a sense for the implementation details we've got in mind at this point. Same ideas though.
+
+.. image:: oep-0004/authz_relationships.jpg
+  :alt: A sketch of the proposed data model that connects service users to
+        resources through applications, scopes, and resource actions.
+
+Governance
+----------
+* Centralize the complete list of scopes but decentralize the associated
+  metadata and definitions
+
+* Centralize application and token management but decentralize token
+  enforcement
+
+* Centralize common implementation patterns as we encounter them (logging,
+  enforcement, etc)
+
+* Access tokens must be reasonably well-contained `JSON Web Tokens`_, signed by
+  the central authorization service, that contain scope information
+
+* Scopes should cover all RESTful web API endpoints [#REST]_ within a service,
+  and may also cover other endpoints as desired by the service team
+
+* Resources should not leak implementation details (such as ``otto_cart``) or,
+  on the other hand, be too broad to be useful (such as ``courseware``).
+
+.. _`JSON Web Tokens`: https://jwt.io
+
+
+Open questions
+--------------
+* Where, exactly, are we going to maintain the list of all scopes, and which
+  scopes should we start with?
```

Comment:
Got it. I'd like to leave that out of the OEP just because it's an implementation detail, while the overall governance note addresses the fact that we need _something_ for developers to use. I expect long term we'll end up with a run-time way to get relevant scopes, but that feels like something we can work towards over time. The minimum baseline I will suggest for implementation is that the list of relevant scopes lives side by side with the Swagger documentation for the API, so that at least it's machine-parseable and in a well-known spec. And the more we can push it into the runtime environment over time, the better.